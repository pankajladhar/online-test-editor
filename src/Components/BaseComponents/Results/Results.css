.Results .Results__TopPanel {
  text-align: center; }
  .Results .Results__TopPanel span {
    display: block;
    padding-bottom: 10px; }
    .Results .Results__TopPanel span.Results__TopPanel__CorrectCount {
      font-size: 20px; }
    .Results .Results__TopPanel span.Results__TopPanel__Percentage {
      font-size: 30px;
      line-height: 1;
      color: #349284; }
