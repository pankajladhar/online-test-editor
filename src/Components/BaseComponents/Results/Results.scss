.Results {
  .Results__TopPanel {
    text-align: center;
    span {
      display: block;
      padding-bottom: 10px;

      &.Results__TopPanel__CorrectCount {
        font-size: 20px;
      }

      &.Results__TopPanel__Percentage {
        font-size: 30px;
        line-height: 1;
        color: #349284;
      }
    }
  }
}