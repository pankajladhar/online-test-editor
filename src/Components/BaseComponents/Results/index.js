import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import Button from './../Button';
import ResultList from './../ResultList';

import './Results.css'

class Results extends PureComponent {
    constructor(props) {
        super(props);
        this.state = {
            showAnswers: false
        }
        this.getCorrectAnswerCount = this.getCorrectAnswerCount.bind(this);
        this.handleClick = this.handleClick.bind(this);
        this.getWrongAnswer = this.getWrongAnswer.bind(this);
    }

    getCorrectAnswerCount(answers) {
        let correctAnswerCount = 0
        let resultListArray = [];
        answers.forEach(element => {
            if (element.correctAnswer === element.answer) {
                correctAnswerCount = correctAnswerCount + 1;
            }
        });
        return correctAnswerCount;
    }

    getWrongAnswer(items) {
        return items.filter((item) => {
            return item.answer !== item.correctAnswer
        })
    }

    handleClick() {
        this.setState({
            showAnswers: !this.state.showAnswers
        })
    }

    render() {
        return (
            <div className="Results">
                <div className="Results__TopPanel">
                    <span className="Results__TopPanel__CorrectCount">
                        {this.getCorrectAnswerCount(this.props.userAnswers)} out of {this.props.noOfques} are correct answer
                    </span>
                    <span className="Results__TopPanel__Percentage">
                        {(this.getCorrectAnswerCount(this.props.userAnswers) / this.props.noOfques) * 100} %
                    </span>
                    <Button className="btn"
                        onClick={this.handleClick}
                        value="Get Correct Answer" />
                </div>
                {
                    this.state.showAnswers && <div className="ResultList__Wrapper">
                        {
                            this.getWrongAnswer(this.props.userAnswers).map((item ,index) => {
                                return (
                                    <ResultList question={item.question}
                                        key={`ResultList-${index}`}
                                        userAnswer={item.answer}
                                        correctAnswer={item.correctAnswer} />
                                )
                            })
                        }
                    </div>
                }
            </div>
        );
    }
}

Results.propTypes = {
    noOfques: PropTypes.number.isRequired,
    userAnswers: PropTypes.arrayOf(PropTypes.shape({
        question: PropTypes.string.isRequired,
        answer: PropTypes.string.isRequired,
        correctAnswer: PropTypes.string.isRequired,
    }))
};

export default Results;