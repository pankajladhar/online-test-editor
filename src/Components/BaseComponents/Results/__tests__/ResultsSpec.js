import React from 'react';
import renderer from 'react-test-renderer';
import { shallow, render } from 'enzyme';
import Results from './../index';

describe('Results', () => {
    let resultsProps = {}
    beforeEach(() => {
        resultsProps = {
            noOfques: 2,
            userAnswers: [
                {
                    question: "Question 1",
                    answer: "Answer 1",
                    correctAnswer: "Answer 1",
                },
                {
                    question: "Question 2",
                    answer: "Answer 2",
                    correctAnswer: "Correct Answer 2",
                }
            ]
        }
    });

    it('renders correctly', () => {
        let tree = renderer.create(
            <Results {...resultsProps} />
        ).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('renders correctly with result list', () => {
        let component = renderer.create(
            <Results {...resultsProps} />
        );
        const instance = component.getInstance()
        instance.handleClick(); 
        expect(component.toJSON()).toMatchSnapshot();
    });

    it('handleClick method should toggle showAnswers state', () =>{
        let resultsInstance = shallow(<Results {...resultsProps} />).instance();
        resultsInstance.handleClick();
        expect(resultsInstance.state.showAnswers).toBeTruthy();

        resultsInstance.handleClick();
        expect(resultsInstance.state.showAnswers).toBeFalsy();
    });

    it('getCorrectAnswerCount method should give correct answers count', () =>{
        let resultsInstance = shallow(<Results {...resultsProps} />).instance();
        expect(resultsInstance.getCorrectAnswerCount(resultsProps.userAnswers)).toBe(1);
    });

    it('getWrongAnswer should reurn worng answers array', () => {
        let resultsInstance = shallow(<Results {...resultsProps} />).instance();
        let worngAnswers = resultsInstance.getWrongAnswer(resultsProps.userAnswers);
        expect(worngAnswers.length).toBe(1);
        expect(worngAnswers[0].answer === worngAnswers[0].correctAnswer).toBeFalsy();
        expect(JSON.stringify(worngAnswers[0])).toBe(JSON.stringify(worngAnswers[0]));
    })

});
