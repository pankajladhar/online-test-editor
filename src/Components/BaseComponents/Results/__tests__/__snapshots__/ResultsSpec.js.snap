// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Results renders correctly 1`] = `
<div
  className="Results"
>
  <div
    className="Results__TopPanel"
  >
    <span
      className="Results__TopPanel__CorrectCount"
    >
      1
       out of 
      2
       are correct answer
    </span>
    <span
      className="Results__TopPanel__Percentage"
    >
      50
       %
    </span>
    <button
      className="Button btn"
      disabled={undefined}
      onClick={[Function]}
      type="button"
    >
      Get Correct Answer
    </button>
  </div>
</div>
`;

exports[`Results renders correctly with result list 1`] = `
<div
  className="Results"
>
  <div
    className="Results__TopPanel"
  >
    <span
      className="Results__TopPanel__CorrectCount"
    >
      1
       out of 
      2
       are correct answer
    </span>
    <span
      className="Results__TopPanel__Percentage"
    >
      50
       %
    </span>
    <button
      className="Button btn"
      disabled={undefined}
      onClick={[Function]}
      type="button"
    >
      Get Correct Answer
    </button>
  </div>
  <div
    className="ResultList__Wrapper"
  >
    <div
      className="ResultList"
    >
      <h2>
        Q. 
        Question 2
      </h2>
      <div
        className="ResultList__Item"
      >
        <span>
          You Answered
        </span>
        <span
          className="ResultList__Item--Wrong"
        >
          Answer 2
        </span>
      </div>
      <div
        className="ResultList__Item"
      >
        <span>
          Correct Answer
        </span>
        <span
          className="ResultList__Item--Correct"
        >
          Correct Answer 2
        </span>
      </div>
    </div>
  </div>
</div>
`;
